NullPointerException in ProjectRepository.convertToGcpProjects

![image](https://cloud.githubusercontent.com/assets/1005544/22937386/7381dbb2-f2a6-11e6-9d5b-2ab39f61fa08.png)

I believe this happened when trying to deploy while logged into an account that does not have GCP access. I forget the exact name, but Google Apps for Your Domain account or some such?

```
java.lang.NullPointerException
	at com.google.cloud.tools.eclipse.projectselector.ProjectRepository.convertToGcpProjects(ProjectRepository.java:93)
	at com.google.cloud.tools.eclipse.projectselector.ProjectRepository.getProjects(ProjectRepository.java:54)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.StandardDeployPreferencesPanel.retrieveProjects(StandardDeployPreferencesPanel.java:365)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.StandardDeployPreferencesPanel.access$3(StandardDeployPreferencesPanel.java:360)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.StandardDeployPreferencesPanel$2.run(StandardDeployPreferencesPanel.java:281)
	at com.google.cloud.tools.eclipse.login.ui.AccountSelector.fireSelectionListeners(AccountSelector.java:147)
	at com.google.cloud.tools.eclipse.login.ui.AccountSelector.selectAccount(AccountSelector.java:133)
	at com.google.cloud.tools.eclipse.login.ui.AccountSelectorObservableValue.doSetValue(AccountSelectorObservableValue.java:59)
	at org.eclipse.core.databinding.observable.value.AbstractObservableValue.setValue(AbstractObservableValue.java:66)
	at org.eclipse.core.databinding.UpdateValueStrategy.doSet(UpdateValueStrategy.java:469)
	at org.eclipse.core.databinding.ValueBinding$3$1.run(ValueBinding.java:206)
	at org.eclipse.core.databinding.observable.Realm$1.run(Realm.java:149)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.databinding.observable.Realm.safeRun(Realm.java:153)
	at org.eclipse.core.databinding.observable.Realm.exec(Realm.java:171)
	at org.eclipse.core.databinding.ValueBinding$3.run(ValueBinding.java:196)
	at org.eclipse.core.databinding.observable.Realm$1.run(Realm.java:149)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.databinding.observable.Realm.safeRun(Realm.java:153)
	at org.eclipse.core.databinding.observable.Realm.exec(Realm.java:171)
	at org.eclipse.core.databinding.ValueBinding.doUpdate(ValueBinding.java:158)
	at org.eclipse.core.databinding.ValueBinding.updateModelToTarget(ValueBinding.java:123)
	at org.eclipse.core.databinding.DataBindingContext.updateTargets(DataBindingContext.java:464)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.DeployPreferencesDialog.createContents(DeployPreferencesDialog.java:82)
	at org.eclipse.jface.window.Window.create(Window.java:426)
	at org.eclipse.jface.dialogs.Dialog.create(Dialog.java:1095)
	at org.eclipse.jface.window.Window.open(Window.java:783)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.standard.StandardDeployCommandHandler.execute(StandardDeployCommandHandler.java:91)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:295)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:55)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:282)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:264)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:132)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:152)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:494)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:488)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:210)
	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.executeItem(HandledContributionItem.java:433)
	at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem.handleWidgetSelection(AbstractContributionItem.java:454)
	at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem$3.handleEvent(AbstractContributionItem.java:482)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4256)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1501)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1509)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1313)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4080)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3706)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1121)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1022)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:150)
	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:693)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:610)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:138)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:673)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:610)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1519)
```

Fixing this causes the Deploy button to enable with no projects:

org.eclipse.e4.core.di.InjectionException: org.eclipse.core.commands.ExecutionException: Project ID must be set for deploy.
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:68)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:252)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:234)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:132)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:152)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:493)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:486)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:210)
	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.executeItem(HandledContributionItem.java:799)
	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.handleWidgetSelection(HandledContributionItem.java:675)
	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.access$7(HandledContributionItem.java:659)
	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem$4.handleEvent(HandledContributionItem.java:592)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4230)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1491)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1514)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1499)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1299)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4072)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3698)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1127)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:337)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1018)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:156)
	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:694)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:337)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:606)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:139)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:380)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:235)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:669)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:608)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1515)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1488)
Caused by: org.eclipse.core.commands.ExecutionException: Project ID must be set for deploy.
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.standard.StandardDeployCommandHandler.getDeployConfiguration(StandardDeployCommandHandler.java:160)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.standard.StandardDeployCommandHandler.launchDeployJob(StandardDeployCommandHandler.java:116)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.standard.StandardDeployCommandHandler.execute(StandardDeployCommandHandler.java:92)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:295)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:56)
	... 41 more

--- a/plugins/com.google.cloud.tools.eclipse.appengine.deploy.test/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.deploy.test/META-INF/MANIFEST.MF
@@ -11,6 +11,7 @@ Require-Bundle: org.hamcrest;bundle-version="1.1.0",
  org.junit;bundle-version="4.12.0"
 Import-Package: com.google.cloud.tools.eclipse.appengine.deploy.standard,
  com.google.cloud.tools.eclipse.test.util,
+ com.google.cloud.tools.eclipse.test.util.project,
  org.eclipse.core.expressions,
  org.mockito;provider=google;version="1.10.19",
  org.mockito.runners;provider=google;version="1.10.19",

--- a/plugins/com.google.cloud.tools.eclipse.appengine.deploy.test/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployPreferencesConverterTest.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.deploy.test/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployPreferencesConverterTest.java
@@ -40,99 +40,125 @@ public class StandardDeployPreferencesConverterTest {
 
   @Test
   public void testToDeployConfiguration_projectId() {
-    when(preferences.get(eq(StandardDeployPreferences.PREF_PROJECT_ID), anyString())).thenReturn("projectid");
+    when(preferences.get(eq(StandardDeployPreferences.PREF_PROJECT_ID), anyString()))
+        .thenReturn("projectid");
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertThat(configuration.getProject(), is("projectid"));
   }
 
   @Test
   public void testToDeployConfiguration_bucketNameIsNull() {
-    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_BUCKET), anyString())).thenReturn(null);
+    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_BUCKET), anyString()))
+        .thenReturn(null);
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertNull(configuration.getBucket());
   }
 
   @Test
   public void testToDeployConfiguration_bucketNameIsEmpty() {
-    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_BUCKET), anyString())).thenReturn("");
+    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_BUCKET), anyString()))
+        .thenReturn("");
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertNull(configuration.getBucket());
   }
 
   @Test
   public void testToDeployConfiguration_bucketNameContainsProtocol() {
-    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_BUCKET), anyString())).thenReturn("gs://bucket");
+    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_BUCKET), anyString()))
+        .thenReturn("gs://bucket");
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertThat(configuration.getBucket(), is("gs://bucket"));
   }
 
   @Test
   public void testToDeployConfiguration_bucketNameDoesNotContainProtocol() {
-    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_BUCKET), anyString())).thenReturn("bucket");
+    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_BUCKET), anyString()))
+        .thenReturn("bucket");
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertThat(configuration.getBucket(), is("gs://bucket"));
   }
 
   @Test
   public void testToDeployConfiguration_promote() {
-    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_ENABLE_AUTO_PROMOTE), anyBoolean())).thenReturn(true);
+    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_ENABLE_AUTO_PROMOTE),
+        anyBoolean())).thenReturn(true);
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertTrue(configuration.getPromote());
   }
 
   @Test
   public void testToDeployConfiguration_promoteNotSetStopPreviousVersionIsUnset() {
-    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_STOP_PREVIOUS_VERSION), anyBoolean())).thenReturn(true);
+    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_STOP_PREVIOUS_VERSION),
+        anyBoolean())).thenReturn(true);
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertNull(configuration.getStopPreviousVersion());
   }
 
   @Test
   public void testToDeployConfiguration_promoteSetStopPreviousVersionIsSet() {
-    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_ENABLE_AUTO_PROMOTE), anyBoolean())).thenReturn(true);
-    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_STOP_PREVIOUS_VERSION), anyBoolean())).thenReturn(true);
+    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_ENABLE_AUTO_PROMOTE),
+        anyBoolean())).thenReturn(true);
+    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_STOP_PREVIOUS_VERSION),
+        anyBoolean())).thenReturn(true);
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertTrue(configuration.getStopPreviousVersion());
   }
 
   @Test
   public void testToDeployConfiguration_promoteSetStopPreviousVersionIsUnset() {
-    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_ENABLE_AUTO_PROMOTE), anyBoolean())).thenReturn(true);
-    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_STOP_PREVIOUS_VERSION), anyBoolean())).thenReturn(false);
+    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_ENABLE_AUTO_PROMOTE),
+        anyBoolean())).thenReturn(true);
+    when(preferences.getBoolean(eq(StandardDeployPreferences.PREF_STOP_PREVIOUS_VERSION),
+        anyBoolean())).thenReturn(false);
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertFalse(configuration.getStopPreviousVersion());
   }
 
   @Test
   public void testToDeployConfiguration_version() {
-    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_VERSION), anyString())).thenReturn("version");
+    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_VERSION), anyString()))
+        .thenReturn("version");
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertThat(configuration.getVersion(), is("version"));
   }
 
   @Test
   public void testToDeployConfiguration_versionIsNull() {
-    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_VERSION), anyString())).thenReturn(null);
+    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_VERSION), anyString()))
+        .thenReturn(null);
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertNull(configuration.getVersion());
   }
 
   @Test
   public void testToDeployConfiguration_versionIsEmpty() {
-    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_VERSION), anyString())).thenReturn("");
+    when(preferences.get(eq(StandardDeployPreferences.PREF_CUSTOM_VERSION), anyString()))
+        .thenReturn("");
     DefaultDeployConfiguration configuration =
-        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences)).toDeployConfiguration();
+        new StandardDeployPreferencesConverter(new StandardDeployPreferences(preferences))
+            .toDeployConfiguration();
     assertNull(configuration.getVersion());
   }
 }

--- a/plugins/com.google.cloud.tools.eclipse.appengine.deploy.test/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployPreferencesTest.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.deploy.test/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployPreferencesTest.java
@@ -17,16 +17,34 @@
 package com.google.cloud.tools.eclipse.appengine.deploy.standard;
 
 import static org.hamcrest.text.IsEmptyString.isEmptyString;
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 
+import org.eclipse.core.resources.IProject;
+import org.junit.Rule;
 import org.junit.Test;
+
+import com.google.cloud.tools.eclipse.test.util.project.TestProjectCreator;
 
 public class StandardDeployPreferencesTest {
 
+  @Rule public final TestProjectCreator projectCreator = new TestProjectCreator();
+  
   @Test
   public void testDefaultProjectId() {
     assertThat(StandardDeployPreferences.DEFAULT.getProjectId(), isEmptyString());
+  }
+  
+  @Test
+  public void testSetProjectId() {
+    IProject project = projectCreator.getProject();
+    StandardDeployPreferences preferences = new StandardDeployPreferences(project);
+    assertThat(preferences.getProjectId(), isEmptyString());
+    preferences.setProjectId("someproject32");
+    assertEquals("someproject32", preferences.getProjectId());
+    preferences.setProjectId(null);
+    assertThat(preferences.getProjectId(), isEmptyString());
   }
 
   @Test

--- a/plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanelTest.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanelTest.java
@@ -17,6 +17,7 @@
 package com.google.cloud.tools.eclipse.appengine.deploy.ui;
 
 import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
@@ -80,6 +81,13 @@ public class StandardDeployPreferencesPanelTest {
   }
 
   @Test
+  public void testHasSelection() {
+    StandardDeployPreferencesPanel deployPanel = new StandardDeployPreferencesPanel(
+        parent, project, loginService, layoutChangedHandler, true, projectRepository);
+    assertFalse(deployPanel.hasSelection());
+  }
+  
+  @Test
   public void testSelectSingleAccount() {
     when(loginService.getAccounts()).thenReturn(new HashSet<>(Arrays.asList(account1)));
     StandardDeployPreferencesPanel deployPanel = new StandardDeployPreferencesPanel(
@@ -119,7 +127,8 @@ public class StandardDeployPreferencesPanelTest {
 
   @Test
   public void testProjectSavedInPreferencesSelected() throws ProjectRepositoryException {
-    IEclipsePreferences node = new ProjectScope(project).getNode(StandardDeployPreferences.PREFERENCE_STORE_QUALIFIER);
+    IEclipsePreferences node =
+        new ProjectScope(project).getNode(StandardDeployPreferences.PREFERENCE_STORE_QUALIFIER);
     node.put("project.id", "projectId1");
     node.put("account.email", EMAIL_1);
     initializeProjectRepository(projectRepository);
@@ -139,13 +148,16 @@ public class StandardDeployPreferencesPanelTest {
     fail("Did not find ProjectSelector widget");
   }
 
-  private void initializeProjectRepository(ProjectRepository projectRepository) throws ProjectRepositoryException {
+  private void initializeProjectRepository(ProjectRepository projectRepository) 
+      throws ProjectRepositoryException {
     GcpProject project1 = new GcpProject("Project1", "projectId1");
     GcpProject project2 = new GcpProject("Project2", "projectId2");
     when(projectRepository.getProjects(any(Credential.class)))
       .thenReturn(Arrays.asList(project1, project2));
-    when(projectRepository.getProject(any(Credential.class), eq("projectId1"))).thenReturn(project1);
-    when(projectRepository.getProject(any(Credential.class), eq("projectId2"))).thenReturn(project2);
+    when(projectRepository.getProject(any(Credential.class), eq("projectId1")))
+        .thenReturn(project1);
+    when(projectRepository.getProject(any(Credential.class), eq("projectId2")))
+        .thenReturn(project2);
   }
 
   private IStatus getAccountSelectorValidationStatus(StandardDeployPreferencesPanel deployPanel) {

--- a/plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java
@@ -76,7 +76,9 @@ public class DeployPreferencesDialog extends TitleAreaDialog {
       setTitleImage(titleImage);
     }
 
-    getButton(IDialogConstants.OK_ID).setText(Messages.getString("deploy"));
+    Button deployButton = getButton(IDialogConstants.OK_ID);
+    deployButton.setText(Messages.getString("deploy"));
+    deployButton.setEnabled(false);
 
     // TitleAreaDialogSupport does not validate initially, let's trigger validation this way
     content.getDataBindingContext().updateTargets();
@@ -107,7 +109,7 @@ public class DeployPreferencesDialog extends TitleAreaDialog {
           @Override
           public int getMessageType(ValidationStatusProvider statusProvider) {
             int type = super.getMessageType(statusProvider);
-            setValid(type != IMessageProvider.ERROR);
+            setValid(type != IMessageProvider.ERROR && content.hasSelection());
             return type;
           }
         });
@@ -176,9 +178,9 @@ public class DeployPreferencesDialog extends TitleAreaDialog {
   }
 
   private void setValid(boolean isValid) {
-    Button okButton = getButton(IDialogConstants.OK_ID);
-    if (okButton != null) {
-      okButton.setEnabled(isValid);
+    Button deployButton = getButton(IDialogConstants.OK_ID);
+    if (deployButton != null) {
+      deployButton.setEnabled(isValid && content.hasSelection());
     }
   }
 

--- a/plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java
@@ -489,4 +489,8 @@ public class StandardDeployPreferencesPanel extends DeployPreferencesPanel {
     expandableComposite.setFont(font);
     FontUtil.convertFontToBold(expandableComposite);
   }
+
+  boolean hasSelection() {
+    return projectSelector.hasSelection();
+  }
 }

--- a/plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployPreferences.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployPreferences.java
@@ -17,6 +17,8 @@
 package com.google.cloud.tools.eclipse.appengine.deploy.standard;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+
 import org.eclipse.core.resources.IProject;
 import org.eclipse.core.resources.ProjectScope;
 import org.eclipse.core.runtime.preferences.IEclipsePreferences;
@@ -61,7 +63,7 @@ public class StandardDeployPreferences {
   }
 
   public void setAccountEmail(String accountEmail) {
-    preferenceStore.put(PREF_ACCOUNT_EMAIL, accountEmail);
+    preferenceStore.put(PREF_ACCOUNT_EMAIL, Strings.nullToEmpty(accountEmail));
   }
 
   public String getProjectId() {
@@ -69,10 +71,7 @@ public class StandardDeployPreferences {
   }
 
   public void setProjectId(String projectId) {
-    if (projectId == null) {
-      projectId = "";
-    }
-    preferenceStore.put(PREF_PROJECT_ID, projectId);
+    preferenceStore.put(PREF_PROJECT_ID, Strings.nullToEmpty(projectId));
   }
 
   public String getVersion() {
@@ -81,7 +80,7 @@ public class StandardDeployPreferences {
   }
 
   public void setVersion(String version) {
-    preferenceStore.put(PREF_CUSTOM_VERSION, version);
+    preferenceStore.put(PREF_CUSTOM_VERSION, Strings.nullToEmpty(version));
   }
 
   public boolean isAutoPromote() {
@@ -99,7 +98,7 @@ public class StandardDeployPreferences {
   }
 
   public void setBucket(String bucket) {
-    preferenceStore.put(PREF_CUSTOM_BUCKET, bucket);
+    preferenceStore.put(PREF_CUSTOM_BUCKET, Strings.nullToEmpty(bucket));
   }
 
   public boolean isStopPreviousVersion() {

--- a/plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployPreferences.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployPreferences.java
@@ -35,6 +35,7 @@ public class StandardDeployPreferences {
   static final String PREF_STOP_PREVIOUS_VERSION = "project.previousVersion.stop";
 
   private IEclipsePreferences preferenceStore;
+  // todo since this is mutable it shouldn;t be in all caps
   public static final StandardDeployPreferences DEFAULT;
 
   static {
@@ -68,6 +69,9 @@ public class StandardDeployPreferences {
   }
 
   public void setProjectId(String projectId) {
+    if (projectId == null) {
+      projectId = "";
+    }
     preferenceStore.put(PREF_PROJECT_ID, projectId);
   }
 

--- a/plugins/com.google.cloud.tools.eclipse.projectselector.test/src/com/google/cloud/tools/eclipse/projectselector/ProjectRepositoryTest.java
+++ b/plugins/com.google.cloud.tools.eclipse.projectselector.test/src/com/google/cloud/tools/eclipse/projectselector/ProjectRepositoryTest.java
@@ -1,0 +1,55 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.eclipse.projectselector;
+
+import java.util.List;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import com.google.api.client.auth.oauth2.Credential;
+
+public class ProjectRepositoryTest {
+
+  @Test
+  public void testGetProjects_nullCredential() throws ProjectRepositoryException {
+    ProjectRepository repository = new ProjectRepository();
+    Credential credential = null;
+    List<GcpProject> projects = repository.getProjects(credential);
+    Assert.assertTrue(projects.isEmpty());
+  }
+
+  @Test
+  public void testGetProjects_credentialWithoutAccess() {
+    ProjectRepository repository = new ProjectRepository();
+    Credential credential = Mockito.mock(Credential.class);
+    try {
+      repository.getProjects(credential);
+      Assert.fail();
+    } catch (ProjectRepositoryException ex) {
+      Assert.assertTrue(ex.getMessage().contains("401"));
+    }
+  }
+  
+  @Test
+  public void testConvertToGcpProjects_null() {
+    List<GcpProject> projects = ProjectRepository.convertToGcpProjects(null);
+    Assert.assertTrue(projects.isEmpty());
+  }
+
+}

--- a/plugins/com.google.cloud.tools.eclipse.projectselector.test/src/com/google/cloud/tools/eclipse/projectselector/ProjectSelectorTest.java
+++ b/plugins/com.google.cloud.tools.eclipse.projectselector.test/src/com/google/cloud/tools/eclipse/projectselector/ProjectSelectorTest.java
@@ -18,6 +18,8 @@ package com.google.cloud.tools.eclipse.projectselector;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
 
 import com.google.cloud.tools.eclipse.test.util.ui.ShellTestResource;
 import java.util.Arrays;
@@ -65,11 +67,13 @@ public class ProjectSelectorTest {
     
     ProjectSelector projectSelector = new ProjectSelector(shellResource.getShell());
     projectSelector.setProjects(projects);
+    assertFalse(projectSelector.hasSelection());
     projectSelector.getViewer().setSelection(new StructuredSelection(selectedProject));
     projectSelector.setProjects(projects.subList(2, projects.size()));
 
     IStructuredSelection selection = projectSelector.getViewer().getStructuredSelection();
     assertThat(selection.size(), is(1));
+    assertTrue(projectSelector.hasSelection());
     assertThat((GcpProject) selection.getFirstElement(), is(selectedProject));
   }
 

--- a/plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectRepository.java
+++ b/plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectRepository.java
@@ -26,6 +26,7 @@ import com.google.api.services.cloudresourcemanager.CloudResourceManager.Project
 import com.google.api.services.cloudresourcemanager.model.ListProjectsResponse;
 import com.google.api.services.cloudresourcemanager.model.Project;
 import com.google.cloud.tools.eclipse.util.CloudToolsInfo;
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Strings;
 import java.io.IOException;
 import java.util.ArrayList;
@@ -62,11 +63,11 @@ public class ProjectRepository {
 
   /**
    * @return a project if the projectId identifies an existing project and the account identified by
-   * {@code credential} has access to the project
+   *         {@code credential} has access to the project
    * @throws ProjectRepositoryException if an error happens while communicating with the backend
    */
-  public GcpProject getProject(Credential credential,
-                               String projectId) throws ProjectRepositoryException {
+  public GcpProject getProject(Credential credential, String projectId) 
+      throws ProjectRepositoryException {
     try {
       if (credential != null && !Strings.isNullOrEmpty(projectId)) {
         return convertToGcpProject(getProjectsApi(credential).get(projectId).execute());
@@ -78,7 +79,7 @@ public class ProjectRepository {
     }
   }
 
-  private Projects getProjectsApi(Credential credential) {
+  private static Projects getProjectsApi(Credential credential) {
     JsonFactory jsonFactory = new JacksonFactory();
     HttpTransport transport = new NetHttpTransport();
     CloudResourceManager resourceManager =
@@ -88,17 +89,20 @@ public class ProjectRepository {
     return projects;
   }
 
-  private List<GcpProject> convertToGcpProjects(List<Project> projects) {
+  @VisibleForTesting
+  static List<GcpProject> convertToGcpProjects(List<Project> projects) {
     List<GcpProject> gcpProjects = new ArrayList<>();
-    for (Project project : projects) {
-      if (!PROJECT_DELETE_REQUESTED.equals(project.getLifecycleState())) {
-        gcpProjects.add(convertToGcpProject(project));
+    if (projects != null) {
+      for (Project project : projects) {
+        if (!PROJECT_DELETE_REQUESTED.equals(project.getLifecycleState())) {
+          gcpProjects.add(convertToGcpProject(project));
+        }
       }
     }
     return gcpProjects;
   }
 
-  private GcpProject convertToGcpProject(Project project) {
+  private static GcpProject convertToGcpProject(Project project) {
     return new GcpProject(project.getName(), project.getProjectId());
   }
 

--- a/plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectSelector.java
+++ b/plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectSelector.java
@@ -73,4 +73,9 @@ public class ProjectSelector extends Composite {
     }
     tableViewer.setSelection(selection);
   }
+  
+  public boolean hasSelection() {
+    ISelection selection = tableViewer.getSelection();
+    return !selection.isEmpty();
+  }
 }

